dorado/particle_track.py

```diff
@@ -903,10 +903,151 @@ def exposure_time(walk_data,
+    visit_freq = np.zeros(raster_size)
+    for ii in list(range(len(walk_data['xinds']))):
+        for jj in list(range(len(walk_data['xinds'][ii]))):
+            visit_freq[walk_data['xinds'][ii][jj],
+                       walk_data['yinds'][ii][jj]] += 1
+
+    # Clip out highest percentile to correct possible outliers
+    vmax = float(np.nanpercentile(visit_freq, clip))
+    visit_freq = np.clip(visit_freq, 0, vmax)
+
+    # If applicable, do smoothing
+    if sigma > 0:
+        visit_freq = gaussian_filter(visit_freq, sigma=sigma)
+        visit_freq[visit_freq==np.min(visit_freq)] = np.nan
+    else:
+        visit_freq[visit_freq==0] = np.nan
+
+    # Normalize to 0-1
+    visit_freq = visit_freq/np.nanmax(visit_freq)
+
+    return visit_freq
+
+
+def nourishment_time(walk_data, raster_size, sigma=0.7, clip=99.5):
+    """Determine the nourishment time of a particle injection
+
+    Function will measure the average length of time particles spend in each
+    area of the domain for a given seed location, as indicated by the history
+    of particle travel times in walk_data. Returns a heatmap raster, in
+    which values indicate the average length of time particles tend to remain
+    in each cell (after spatial filtering to reduce noise in the random walk).
+
+    **Inputs** :
+
+        walk_data : `dict`
+            Dictionary of all prior x locations, y locations, and travel
+            times (the output of run_iteration)
+
+        raster_size : `tuple`
+            Tuple (L,W) of the domain dimensions, i.e. the output of
+            numpy.shape(raster).
+
+        sigma : `float`, optional
+            Degree of spatial smoothing of the area, implemented using a
+            Gaussian kernal of the same sigma, via scipy.ndimage.gaussian_filter
+            Default is light smoothing with sigma = 0.7 (to turn off smoothing,
+            set sigma = 0)
+
+        clip : `float`, optional
+            Percentile at which to truncate the distribution. Particles which
+            get stuck can lead to errors at the high-extreme, so this 
+            parameter is used to normalize by a "near-max". Default is the
+            99.5th percentile. To use true max, specify clip = 100
+
+    **Outputs** :
+
+        mean_time : `numpy.ndarray`
+            Array of mean occupation time, with cell values representing the 
+            mean time particles spent in that cell. If sigma > 0, the array
+            values include spatial filtering
+
+    """
+    if sigma > 0:
+        from scipy.ndimage import gaussian_filter
+    
+    # Measure visit frequency
+    visit_freq = np.zeros(raster_size)
+    time_total = visit_freq.copy()
+    for ii in list(range(len(walk_data['xinds']))):
+        for jj in list(range(1, len(walk_data['xinds'][ii])-1)):
+            # Running total of particles in this cell to find average
+            visit_freq[walk_data['xinds'][ii][jj],
+                       walk_data['yinds'][ii][jj]] += 1
```

Comment:
It looks like the calculation for `nourishment_area()` is embedded within the `nourishment_time()` function as well. Do you think there is value to combining these functions into a `calculate_nourishment()` function that has an input parameter "nourishment_type" or something similar that controls whether the visit frequency, mean time, or both are returned by the function? I can imagine wanting both the visit frequencies and mean times and it seems like they could be rolled into a single function.

Reply:
I thought about it and I think it will be simpler to use/read the code if we keep them as two separate functions. I think they're different enough that their use cases aren't guaranteed to overlap. Plus, I think each function may work best with different degrees of smoothing/clipping, so it's easiest if they're not tied together.